api/src/main/java/jakarta/servlet/http/HttpServlet.java

```diff
@@ -598,78 +598,135 @@ public void service(ServletRequest req, ServletResponse res) throws ServletExcep
 
     private static final ResourceBundle lStrings = ResourceBundle.getBundle("jakarta.servlet.http.LocalStrings");
 
-    private NoBodyOutputStream noBody;
+    private NoBodyOutputStream noBodyOutputStream;
     private PrintWriter writer;
-    private boolean didSetContentLength;
+    private long setContentLength = -1;
     private boolean usingOutputStream;
 
     // file private
     NoBodyResponse(HttpServletResponse r) {
         super(r);
-        noBody = new NoBodyOutputStream();
+    }
+
+    public long getSetContentLength() {
+        return setContentLength;
     }
 
     // file private
-    void setContentLength() {
-        if (!didSetContentLength) {
-            if (writer != null) {
+    void checkContentLength() throws IOException {
+        // Set the content length IFF it is not already set and the app the output stream
+        // has been created.
+
+        if (!isCommitted()) {
+            if (setContentLength == -1) {
```

Comment:
What if the user explicitly set `Transfer-Encoding: chunked` in Undertow's case if the user has explicitly set a TE header we won't mess with the content length.

Reply:
Ah yes.
But I think that reveals the folly of this PR.    I don't know how we are going to perfectly mimic the headers that a container might use in GET without actually deferring the handling to the container.
The original idea was to set content-length if a similar GET request would have, but there is no compulsion on a container to set content-length even if it knows it.

Perhaps a better idea would be to provide some way for a container to disable the whole nobody wrapper stuff anyway. That way the container can opt to handle HEAD itself.